M2/Macaulay2/m2/methods.m2

```diff
@@ -516,38 +502,88 @@ dispatcherFunctions = join (dispatcherFunctions, {
-
-addHook = method()
+-- also see hooks in code.m2
+-- TODO: get this to also work with methods like ((_, =), Y, Z) or ((SPACE, =), Function, Thing)
+-- TODO: get this to work with lookup and flagLookup
+
+protect symbol Hooks
+protect symbol HookAlgorithms
+protect symbol HookPriority
+
+getHookStore = (key, create) -> (
+    -- retrieve (or create) the mutable hashtable of Hooks installed on a symbol or hashtable
+    sym := if key#?0 then key#0 else error "createHookStore: encountered empty method key";
+    sym  = if instance(sym, Symbol) then sym else getGlobalSymbol toString sym; -- TODO: is toString the right function?
+    -- this branch depends on whether hooks are installed on a symbol or a hash table
+    if #key == 1 then ( key = sym; if value sym =!= sym then value sym else if create then sym <- new MutableHashTable from {} ) else (
+	obj := youngest drop(key, 1);
+	if instance(obj, MutableHashTable) then ( if       obj.?Hooks then       obj.Hooks else if create then       obj.Hooks = new MutableHashTable from {} ) else
+	if instance(obj, HashTable)        then ( if obj.cache.?Hooks then obj.cache.Hooks else if create then obj.cache.Hooks = new MutableHashTable from {} )))
+
+addHook = method(
+    Options => {
+	Strategy => null
+	-- Priority, Description, ...?
+	}
+    )
+addHook(HashTable, Thing, Function) := opts -> (obj, key, hook) -> addHook((key, obj), hook, opts)
+addHook(Symbol,           Function) := opts -> (sym,      hook) -> addHook(1:sym,      hook, opts)
+addHook(Sequence,         Function) := opts -> (key,      hook) -> (
+    store := getHookStore(key, true);
+    -- this is the hashtable of Hooks for a specific key, which stores HookAlgorithms and HookPriority
+    if not store#?key then store#key = new MutableHashTable from {
+	HookAlgorithms => new MutableHashTable from {},  -- a mutable hash table "strategy key" => "strategy code"
+	HookPriority   => new MutableList      from {}}, -- a mutable list of strategy keys, in order
+    store = store#key;
+    ind := #store.HookPriority; -- index to add the hook in the list; TODO: use Priority to insert in the middle?
+    alg := if opts.Strategy =!= null then opts.Strategy else ind;
+    store.HookPriority#ind = alg;
+    store.HookAlgorithms#alg = hook)
+
+-- FIXME: removing and adding without strategy is buggy
 removeHook = method()
-runHooks = method(Options => true)
+removeHook(HashTable, Thing, Thing) := (obj, key, alg) -> removeHook((key, obj), alg)
+removeHook(Symbol,           Thing) := (sym,      alg) -> removeHook(1:sym,      alg)
+removeHook(Sequence,         Thing) := (key,      alg) -> (
+    store := getHookStore(key, false);
+    store  = if store =!= null and store#?key then store#key else return;
+    store.HookPriority = delete(alg, store.HookPriority);
+    remove(store.HookAlgorithms, alg))
+
+-- tracking debugInfo
+infoLevel    := -1
+pushInfoLevel =  n     -> (infoLevel = infoLevel + n; n)
+popInfoLevel  = (n, s) -> (infoLevel = infoLevel - n; s)
```

Comment:
There is; it makes the code simpler. Try this:
```m2
level = -1
pushLevel :=  n     -> (level = level + n; n)
popLevel  := (n, s) -> (level = level - n; s)
g = () -> print level
f = n -> popLevel(pushLevel 1, if n > 0 then f(n-1) else g())
f 10
```

Reply:
You will confuse the user reading it.  And this is shorter and easier to read
```m2
level = -1
pushLevel :=  n -> level = level + n
popLevel  :=  n -> level = level - n
g = () -> print level
f = n -> (pushLevel 1; first (if n > 0 then f(n-1) else g(), popLevel 1))
f 10
```

Reply:
The code is trivial to read, and I learned this trick in a programming class so I don't think anyone will be confused. I also like the brevity of it.

Reply:
Well, if you like brevity, you'll love this better:
```m2
i1 : level = -1;

i2 : levelWrap = (n,f) -> (level = level + n; r := f(); level = level - n; r);

i3 : g = () -> (print level;foo)

o3 = g

o3 : FunctionClosure

i4 : f = n -> levelWrap(1,() -> if n > 0 then f(n-1) else g());

i5 : level,f 10,level
10

o5 = (-1, foo, -1)

o5 : Sequence
```

Reply:
I don't really. I think you misunderstood the point of my construction, which is my fault, because I didn't give a good enough example. Consider these three:
```m2
i1 : level = -1;

i2 : pushLevel :=  n     -> (level = level + n; n);

i3 : popLevel  := (n, s) -> (level = level - n; s);

i4 : levelWrap := (n, f) -> (level = level + n; r := f(); level = level - n; r);

i5 : g = () -> recursionDepth();

i6 : f1 = n -> popLevel(pushLevel 1, if n > 0 then f1(n-1) else g());

i7 : f2 = n -> (pushLevel 1; first (if n > 0 then f2(n-1) else g(), popLevel(1,)));

i8 : f3 = n -> levelWrap(1,() -> if n > 0 then f3(n-1) else g());

i9 : (f1(10), f2(10), f3(10))

o9 = (12, 23, 34)
```
My version is strictly the most efficient of all three because of tail call optimization. So I'll keep my code. By the way, I used the same trick here:
https://github.com/Macaulay2/M2/blob/9178cff426ebea23aa17a9f332989c5c4a97f4f4/M2/Macaulay2/m2/html.m2#L65-L68
and used here:
https://github.com/Macaulay2/M2/blob/9178cff426ebea23aa17a9f332989c5c4a97f4f4/M2/Macaulay2/m2/html.m2#L103-L107

Reply:
Just in case I need any more justification for why my version is better:
```m2
i12 : f1(290)

o12 = 292

i13 : f2(290)
stdio:7:12:(3):[301]: error: recursion limit of 300 exceeded

i14 : f3(290)
stdio:8:43:(3):[300]: error: recursion limit of 300 exceeded
```